amethyst_ui/src/bundle.rs

```diff
@@ -59,21 +53,47 @@ where
             &["ui_loader"],
         );
         builder.add(
-            UiKeyboardSystem::new(),
-            "ui_keyboard_system",
-            &["font_processor"],
+            CacheSelectionOrderSystem::<G>::new(),
+            "selection_order_cache",
+            &[],
+        );
+        builder.add(
+            SelectionMouseSystem::<G, A, B>::new(),
+            "ui_mouse_selection",
+            &[],
+        );
+        builder.add(
+            SelectionKeyboardSystem::<G>::new(),
+            "ui_keyboard_selection",
+            // Because when you press tab, you want to override the previously selected elements.
+            &["ui_mouse_selection"],
```

Comment:
Since these can't execute in parallel and they handle roughly the same thing I have to wonder why they're separate systems.

Reply:
Code separation, which is the whole point of this PR ;)

Reply:
Both systems have an input of `winit` events and an output of `Selected` components.  So it doesn't seem like we're separating concerns.  I see two reasons you might want to separate these

1. It's considered important that keyboard selection always overrides mouse selection.  In which case, why?

2. We don't like large match statements, and want to separate them into a mouse section and a keyboard section.  This could be achieved with functions like `handle_keyboard_event(&mut self, event: &Event)` and `handle_mouse_event(&mut self, event: &Event)`.  Personally I don't see the issue with large match statements, as it allows you to observe all of the possible control flows for selection in a single place.

Reply:
Systems don't cost anything. Also you can decide to enable just one of the two (if you want keyboard only and no mouse.) In my head, the keyboard input takes priority over the mouse for selecting stuff. Maybe that's not true, but enforcing the order prevents non-deterministic events. If you run two times the game with input simulation, you need one to be before the other to be deterministic.

Reply:
being able to enable only one of the two actually makes a lot of sense as I can definitely imagine contexts where that might be valuable, so we can keep it.  As an aside

- It's not about run time cost as much as maintenance cost
- There's no reason for keyboard to take priority, only one form of selection is used at a time typically.
- A single system is just as deterministic as two systems if the order of events is the same.  If the order of events is different then the behavior should be different and can no longer be said to be deterministic.

That being said you made a good point about selective handling for device input, so I'll resolve this.

Reply:
It's not about run time cost as much as maintenance cost

To me having small, separated system, lowers the maintenance cost by a lot. (At least from my experience)